Task4.1/DiagonalMatrix.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Task41
+{
+    public class DiagonalMatrix<T>
+    {
+        private readonly T[] diagonalElements;
```

Comment:
Not very good idea to store history and undo method directly in Diagonal matrix. This functionality has to be under an external control-> if necessary it can track, if not- not track. Move changedHistory and Undo to Matrix tracker, in other case no sense in an event